File: Matlab/real_Leja_phi.m
function [polynomial, num_rhs_calls, Leja_convergence] = real_Leja_phi(u, dt, RHS_function, interp_func, c, Gamma, phi_function, rel_tol)
    %%% ---------------------------------------------------
    %
    % Parameters
    % ----------
    % u                       : 1D vector u (input)
    % dt                      : Step size
    % RHS_function            : RHS function
    % interp_func             : function to be multiplied to phi function
    % c                       : Shifting factor
    % Gamma                   : Scaling factor
    % phi_function            : phi function
    % rel_tol                 : Accuracy of the polynomial so formed

    % Returns
    % ----------
    % polynomial              : Polynomial interpolation of 'interp_func' 
    %                           multiplied by 'phi_func' at real Leja points
    % num_rhs_calls           : # of RHS calls
    % Leja_convergence        : 0 - did not converge, 1 - converged
    %
    %%% ---------------------------------------------------

    %%% Phi function applied to 'interp_func' (scaled and shifted)
    function phi_function_array = func(xx)

        zz = (dt * (c + Gamma*xx));
        phi_function_array = phi_function(zz);

    end

    %%% Array of Leja points
    Leja_file = fopen('Leja_points.txt', 'r');
    Leja_X = fscanf(Leja_file, '%f');

    %%% Compute the polynomial coefficients
    coeffs = Divided_Difference(@func, Leja_X);

    %%% ------------------------------------------------------------------- %%%

    %%% a_0 term (form the polynomial)
    poly = interp_func;
    poly = coeffs(1) * poly;

    %%% ------------------------------------------------------------------- %%%

    %%% a_1, a_2 .... a_n terms
    max_Leja_pts = 500;                                      % Max number of Leja points
    y = interp_func;                                         % x values of the polynomial
    epsilon = 1e-7;
    Leja_convergence = 0;                                    % 0 - did not converge, 1 - converged

    %%% Iterate until converges
    for ii = 2 : max_Leja_pts

        %%% Compute the numerical Jacobian
        Jacobian_function = (RHS_function(u + (epsilon * y)) - RHS_function(u))/epsilon;

        %%% Re-scale and re-shift
        y = y * (-c/Gamma - Leja_X(ii - 1));
        y = y + (Jacobian_function/Gamma);

        %%% Approx. error incurred (accuracy)
        poly_error = sqrt(sum(abs(y).^2))/length(y) * abs(coeffs(ii));

        %%% If new number (next order) to be added < tol, break loop
        if  poly_error < rel_tol
%             disp('No. of Leja points used (real phi) = '), ii
%             disp('----------Tolerance reached---------------')
    
            poly = poly + (coeffs(ii) * y);
            Leja_convergence = 1;

            break

        %%% To stop diverging
        elseif poly_error > 1e7
%             disp('Starts to diverge after ", ii, " iterations.')
            polynomial = interp_func;
            num_rhs_calls = 2*ii; 
            Leja_convergence = 0;

            return

        else
           poly = poly + (coeffs(ii) * y);
        end
    end

    %%% ------------------------------------------------------------------- %%%

    %%% Solution
    polynomial = poly;
    num_rhs_calls = 2*ii;

end

File: Python/real_Leja_phi.py
import numpy as np
from Jacobian import Jacobian
from Divided_Difference import Divided_Difference

def real_Leja_phi(u, dt, RHS_function, interp_vector, integrator_coeffs, c, Gamma, Leja_X, phi_function, tol):
    """
    Computes the polynomial interpolation of 'phi_function' applied to 'interp_vector' at real Leja points.


        Parameters
        ----------
        u                       : numpy array
                                    State variable(s)
        dt                      : double
                                    Step size
        RHS_function            : user-defined function 
                                    RHS function
        interp_vector           : numpy array
                                    Vector to be interpolated
        c                       : double
                                    Shifting factor
        Gamma                   : double
                                    Scaling factor
        Leja_X                  : numpy array
                                    Array of Leja points
        phi_function            : function
                                    phi function
        tol                     : double
                                    Accuracy of the polynomial so formed

        Returns
        ----------
        polynomial_array        : numpy array(s)
                                    Polynomial interpolation of 'interp_vector' 
                                    multiplied by 'phi_function' at real Leja points
        ii+1                    : int
                                    # of RHS calls
        convergence             : int
                                    0 -> did not converge, 1 -> converged

    """

    ###? Initialize parameters and arrays
    convergence = 0                                                             #* 0 -> did not converge, 1 -> converged
    num_interpolations = len(integrator_coeffs)                                 #* Number of interpolations in vertical
    max_Leja_pts = len(Leja_X)                                                  #* Max number of Leja points  
    phi_function_array = np.zeros((len(Leja_X), num_interpolations))            #* Phi function applied to 'interp_vector'
    poly_coeffs = np.zeros((len(Leja_X), num_interpolations))                   #* Polynomial coefficients
    polynomial_array = np.zeros((len(interp_vector), num_interpolations))       #* Polynomial array
    y = interp_vector.copy()                                                    #* To avoid changing 'interp_vector'
    
    ###? Loop for vertical implementation
    for ij in range(0, num_interpolations):
        
        ###? Phi function applied to 'interp_vector' (scaled and shifted)
        phi_function_array[:, ij] = phi_function(integrator_coeffs[ij] * dt * (c + Gamma*Leja_X))
        
        ###? Compute polynomial coefficients
        poly_coeffs[:, ij] = Divided_Difference(Leja_X, phi_function_array[:, ij]) 
        
        ###? Form the polynomial: p_0 term
        polynomial_array[:, ij] = interp_vector * poly_coeffs[0, ij]
    
    ###? p_1, p_2, ...., p_n terms; iterate until converges
    for ii in range(1, max_Leja_pts):

        ###? y = y * ((z - c)/Gamma - Leja_X)
        y = (Jacobian(RHS_function, u, y)/Gamma) + (y * (-c/Gamma - Leja_X[ii - 1]))

        ###? Error estimate; poly_error = |coeffs[nn]| ||y||
        poly_error = np.linalg.norm(y) * abs(poly_coeffs[ii, np.argmax(integrator_coeffs)])
        
        ###? Keep adding terms to the polynomial
        for ij in range(0, num_interpolations):

            ### To prevent diverging, restart simulations with smaller dt
            if poly_error > 1e7:
                convergence = 0
                polynomial_array[:, ij] = u
                return polynomial_array, ii+1, convergence

            ###? Add the new term to the polynomial
            polynomial_array[:, ij] = polynomial_array[:, ij] + (poly_coeffs[ii, ij] * y)
            
        ###? If new term to be added < tol, break loop
        if  poly_error < ((tol*np.linalg.norm(polynomial_array)) + tol):
            convergence = 1
            print("Leja points used: ", ii)
            break

        ###! Warning flags
        if ii == max_Leja_pts - 1:
            print("Warning!! Max. # of Leja points reached without convergence!!")
            print("Reduce dt.")
            break

    return polynomial_array, 3*ii, convergence
